clients/python-wrapper/lakefs/object.py

```diff
@@ -84,9 +84,23 @@ def closed(self) -> bool:
 
     def close(self) -> None:
         """
-        Closes the current file descriptor for IO operations
+        Finalizes any existing operations on object and close open descriptors
+        Inheriting classes need to implement _finalize() and _close() respectfully
         """
+        if self._is_closed:
+            return
+
         self._is_closed = True
+        self._finalize()
+        self._close()
```

Comment:
Why do we need more methods to close? method close should handle its resources in each level of the hierarchy and protect from multiple calls.

Reply:
Due to the class inheritence - instead of creating a __enter__, __exit__ method for each class, only the base class implements them.
Then we have to differentiate between completion with failure, and completion without failure